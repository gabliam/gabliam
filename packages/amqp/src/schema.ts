import { Joi } from '@gabliam/core';
import { UNDEFINED_VALUE } from './constants';

const queueOptionsValidator = Joi.object().keys({
  exclusive: Joi.boolean(),
  durable: Joi.boolean(),
  autoDelete: Joi.boolean(),
});

const queueConfigurationValidator = Joi.object().keys({
  queueName: Joi.string()
    .trim()
    .required(),
  options: queueOptionsValidator.default(),
});

const queuesValidor = Joi.object()
  .pattern(/.*/, queueConfigurationValidator)
  .min(1);

const connectionValidator = Joi.object().keys({
  name: Joi.string()
    .trim()
    .default('default'),
  url: Joi.string()
    .trim()
    .required(),
  undefinedValue: Joi.string()
    .trim()
    .default(UNDEFINED_VALUE),
  queues: queuesValidor,
  gzipEnabled: Joi.boolean().default(false),
});

export const configurationValidator = Joi.alternatives(
  connectionValidator,
  Joi.array().items(connectionValidator)
);
