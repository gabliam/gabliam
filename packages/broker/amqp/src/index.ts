import { AmqpPlugin } from './amqp-plugin';

export * from './amqp-connection';
export * from './amqp-manager';
export * from './errors';
export * from './interfaces';
export * from './metadatas';
export * from './queue';
export { AmqpPlugin as default };
