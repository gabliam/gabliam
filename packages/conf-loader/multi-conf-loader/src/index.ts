import {
  configResolver,
  LoaderConfigParseError,
  Resolver,
} from '@gabliam/core';
import d from 'debug';
import fs from 'fs';
import yaml from 'js-yaml';
import _ from 'lodash';
import { inspect, promisify } from 'util';

// Promisify
const glob = promisify(require('glob')); // require and no import for typings bug
const readFile = promisify(fs.readFile);

const debug = d('@gabliam/multi-conf-loader');

async function loadConstant(
  folder: string,
  projectName: string,
  resolver: Resolver,
  profile?: string,
) {
  const files: string[] = await glob(`**/constants?(-+([a-zA-Z])).@(yml)`, {
    cwd: folder,
  });

  let constants: object = {};

  if (!files || files.length === 0) {
    return constants;
  }

  const defaultProfileFile = files.find((file) => file === `constants.yml`);
  if (defaultProfileFile) {
    constants = _.merge(
      {},
      constants,
      await loadFile(`${folder}/${defaultProfileFile}`, resolver),
    );
  }

  if (profile) {
    const profileFile = files.find(
      (file) => file === `constants-${profile}.yml`,
    );

    if (profileFile) {
      constants = _.merge(
        {},
        constants,
        await loadFile(`${folder}/${profileFile}`, resolver),
      );
    }
  }
  return constants;
}

/**
 * FileLoader
 */
const multiConfLoader = async (
  { folder, projectName }: { folder: string; projectName: string },
  profile?: string,
) => {
  debug('loadConfig', folder);
  const resolver = configResolver(folder);
  const constants = await loadConstant(folder, projectName, resolver, profile);

  // create tag config
  const ConfigYamlType = new yaml.Type('!config', {
    kind: 'scalar',
    construct: function (data) {
      return _.get(constants, data, {});
    },
    instanceOf: Object,
  });

  const GAB_SCHEMA = yaml.Schema.create([ConfigYamlType]);

  const files: string[] = await glob(`**/application?(-+([a-zA-Z])).@(yml)`, {
    cwd: folder,
  });

  let config: object = {};

  if (!files || files.length === 0) {
    return config;
  }

  const defaultProfileFile = files.find((file) => file === `application.yml`);
  if (defaultProfileFile) {
    config = _.merge(
      {},
      config,
      await loadConfigProject(
        `${folder}/${defaultProfileFile}`,
        projectName,
        GAB_SCHEMA,
        resolver,
      ),
    );
  }

  if (profile) {
    const profileFile = files.find(
      (file) => file === `application-${profile}.yml`,
    );

    if (profileFile) {
      config = _.merge(
        {},
        config,
        await loadConfigProject(
          `${folder}/${profileFile}`,
          projectName,
          GAB_SCHEMA,
          resolver,
        ),
      );
    }
  }
  debug('loadedConfig', inspect(config, { depth: null }));
  return config;
};

async function loadConfigProject(
  filePath: string,
  projectName: string,
  schema: yaml.Schema,
  resolver: Resolver,
) {
  const data = await readFile(filePath, 'utf8');
  try {
    const vals = yaml.loadAll(data, undefined, { schema }) || [];
    return await resolver(_.find(vals, <any>{ projectName }) || {});
  } catch (e) {
    throw new LoaderConfigParseError(filePath, e);
  }
}

async function loadFile(filePath: string, resolver: Resolver): Promise<Object> {
  const data = await readFile(filePath, 'utf8');
  try {
    return await resolver(yaml.load(data) || {});
  } catch (e) {
    throw new LoaderConfigParseError(filePath, e);
  }
}

export default multiConfLoader;
