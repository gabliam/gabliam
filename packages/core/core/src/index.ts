import 'reflect-metadata';

export {
  decorate,
  inject,
  injectable,
  interfaces as inversifyInterfaces,
  multiInject,
  named,
  optional,
  postConstruct,
  tagged,
  targetName,
  unmanaged,
} from 'inversify';
export * from './common';
export {
  APP_CONFIG,
  CORE_CONFIG,
  INJECT_CONTAINER_KEY,
  VALUE_EXTRACTOR,
} from './constants';
export * from './container';
export * from './decorator';
export * from './errors';
export * from './gabliam';
export * from './gabliam-options';
export * from './gabliam-utils';
export * from './interfaces';
export * from './joi';
export { configResolver, FileLoader, Loader, Resolver } from './loaders';
export * from './metadatas';
export * from './reflection';
export * from './registry';
export * from './value-extractor';
