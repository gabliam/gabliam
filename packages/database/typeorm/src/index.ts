import { TypeOrmPlugin } from './typeorm-plugin';

export { ConnectionManager } from './connection-manager';
export * from './connection-options-reader';
export * from './constant';
export * from './typeorm';
export { TypeOrmPlugin as default };
