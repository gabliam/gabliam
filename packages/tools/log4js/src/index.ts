import { Log4jsPlugin } from './log4js-plugin';

export * from './log4js';

export { Log4jsPlugin as default };
