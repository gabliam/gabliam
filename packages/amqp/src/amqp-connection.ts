import { ValueExtractor } from '@gabliam/core';
import * as amqp from 'amqp-connection-manager';
import { ConfirmChannel, ConsumeMessage, Message } from 'amqplib';
import * as PromiseB from 'bluebird';
import * as _ from 'lodash';
import * as uuid from 'uuid';
import { DEFAULT_PARAM_VALUE, PARAMETER_TYPE } from './constants';
import { AmqpConnectionError, AmqpTimeoutError } from './errors';
import {
  ConsumeConfig,
  ConsumeOptions,
  ConsumerHandler,
  Controller,
  ParameterMetadata,
  RabbitHandlerMetadata,
  SendOptions,
} from './interfaces';
import { Queue } from './queue';
import { log4js } from '@gabliam/log4js';
import { gzip, gunzip } from 'zlib';

enum ConnectionState {
  stopped,

  running,

  starting,

  stopping,
}

/**
 * Amqp Connection
 */
export class AmqpConnection {
  private logger = log4js.getLogger(AmqpConnection.name);

  private state = ConnectionState.stopped;

  private connection: amqp.AmqpConnectionManager;

  private channel: amqp.ChannelWrapper;

  private consumerList: ConsumeConfig[] = [];

  constructor(
    public indexConfig: number,
    public name: string,
    private url: string,
    private undefinedValue: string,
    private queues: Queue[],
    private valueExtractor: ValueExtractor,
    private gzipEnabled: boolean
  ) {}

  /**
   * Start the connection
   */
  async start() {
    if (this.state !== ConnectionState.stopped) {
      return;
    }

    this.state = ConnectionState.starting;
    this.connection = amqp.connect([this.url]);
    this.channel = this.connection.createChannel({
      setup: async (channel: ConfirmChannel) => {
        for (const queue of this.queues) {
          await channel.assertQueue(queue.queueName, queue.queueOptions);
        }
        for (const { queueName, handler, options } of this.consumerList) {
          await channel.consume(queueName, handler, options);
        }
      },
    });

    return new Promise((resolve, reject) => {
      this.connection.once('disconnect', (err: any) => {
        if (_.get(err, 'err.errno', undefined) === 'ENOTFOUND') {
          this.connection.close();
          this.channel.removeAllListeners('connect');
          this.channel.removeAllListeners('error');
          reject(err.err);
        } else {
          this.logger.error(`Amqp error %O`, err);
        }
      });

      this.state = ConnectionState.running;
      const isConnect = () => {
        this.connection.removeAllListeners('disconnect');
        resolve();
      };
      this.channel.once('connect', isConnect);
      this.channel.once('error', isConnect);
    });
  }

  /**
   * Add a consumer for a queue
   */
  addConsume(
    queue: string,
    handler: ConsumerHandler,
    options?: ConsumeOptions
  ) {
    const queueName = this.getQueueName(queue);
    if (!this.queueExist(queueName)) {
      throw new Error(`queue "${queueName}" doesn't exist`);
    }
    this.consumerList.push({ queueName, handler, options });
  }

  /**
   * contrust consumer with controller instance and HandlerMetadata
   */
  constructAndAddConsume(
    handlerMetadata: RabbitHandlerMetadata,
    parameterMetadata: ParameterMetadata[],
    controller: Controller
  ) {
    let consumeHandler: ConsumerHandler;
    if (handlerMetadata.type === 'Listener') {
      consumeHandler = this.constructListener(
        handlerMetadata,
        parameterMetadata,
        controller
      );
    } else {
      consumeHandler = this.constructConsumer(
        handlerMetadata,
        parameterMetadata,
        controller
      );
    }

    this.addConsume(
      handlerMetadata.queue,
      consumeHandler,
      handlerMetadata.consumeOptions
    );
  }

  /**
   * Send a content to a queue.
   * Content can be undefined
   */
  async sendToQueue(queue: string, content: any, options?: SendOptions) {
    const queueName = this.getQueueName(queue);
    const channel = this.getChannel();
    if (channel === null) {
      throw new AmqpConnectionError('Connection error');
    }
    await channel.sendToQueue(queueName, await this.contentToBuffer(content), {
      contentEncoding: this.gzipEnabled ? 'gzip' : undefined,
      contentType: 'application/json',
      ...options,
    });
  }

  /**
   * Send a content to a queue and Ack the message
   * Content can be undefined
   */
  async sendToQueueAck(
    queue: string,
    content: any,
    msg: Message,
    options?: SendOptions
  ) {
    const queueName = this.getQueueName(queue);
    const channel = this.getChannel();
    if (channel === null) {
      throw new AmqpConnectionError('Connection error');
    }
    await channel.sendToQueue(queueName, await this.contentToBuffer(content), {
      contentEncoding: this.gzipEnabled ? 'gzip' : undefined,
      contentType: 'application/json',
      ...options,
    });
    await this.channel.ack(msg);
  }

  /**
   * Basic RPC pattern with conversion.
   * Send a Javascrip object converted to a message to a queue and attempt to receive a response, converting that to a Java object.
   * Implementations will normally set the reply-to header to an exclusive queue and wait up for some time limited by a timeout.
   */
  async sendAndReceive<T = any>(
    queue: string,
    content: any,
    options: SendOptions = {},
    timeout: number = 5000
  ): Promise<T> {
    let onTimeout = false;
    let replyTo: string;
    let chan: ConfirmChannel | null;
    let promise = new PromiseB<T>((resolve, reject) => {
      const queueName = this.getQueueName(queue);
      if (!options.correlationId) {
        options.correlationId = uuid();
      }
      const correlationId = options.correlationId;

      if (!options.replyTo) {
        options.replyTo = `amqpSendAndReceive${uuid()}`;
      }

      if (!options.expiration) {
        options.expiration = '' + timeout;
      }

      replyTo = options.replyTo;
      chan = this.getChannel();
      if (chan === null) {
        return reject(new AmqpConnectionError('Connection error'));
      }
      // create new Queue for get the response
      chan
        .assertQueue(replyTo, {
          exclusive: false,
          autoDelete: true,
          durable: false,
        })
        .then(() => {
          return chan!.consume(replyTo, async (msg: ConsumeMessage | null) => {
            if (msg === null) {
              return reject(new Error('Message is null'));
            }
            if (!onTimeout && msg.properties.correlationId === correlationId) {
              resolve(this.parseContent(msg));
            }
            chan!.ack(msg);
            try {
              await chan!.deleteQueue(replyTo);
            } catch {}
          });
        })
        .then(async () => {
          chan!.sendToQueue(queueName, await this.contentToBuffer(content), {
            contentEncoding: this.gzipEnabled ? 'gzip' : undefined,
            contentType: 'application/json',
            ...options,
          });
        })
        // catch when error amqp (untestable)
        .catch(
          // prettier-ignore
          /* istanbul ignore next */
          async (err: any) => {
            reject(err);
            if (chan) {
              try {
                await chan.deleteQueue(replyTo);
              } catch {}
            }
          }
        );
    });

    if (timeout) {
      promise = promise
        .timeout(timeout)
        .catch(PromiseB.TimeoutError, async e => {
          onTimeout = true;
          if (chan) {
            try {
              await chan.deleteQueue(replyTo);
            } catch {}
          }
          throw new AmqpTimeoutError((<any>e).message);
        });
    }

    return promise;
  }

  /**
   * Stop the connection
   */
  async stop() {
    if (this.state !== ConnectionState.running) {
      return;
    }
    this.state = ConnectionState.stopping;

    try {
      this.channel.removeAllListeners('connect');
      this.channel.removeAllListeners('error');
      await this.connection.close();
    } catch {}

    this.state = ConnectionState.stopped;
  }

  /**
   * Test if queue exist
   */
  queueExist(queueName: string) {
    for (const queue of this.queues) {
      if (queue.queueName === queueName) {
        return true;
      }
    }
    return false;
  }

  /**
   * Get the real queueName
   *
   * Search if the queueName is the index of the map of queues => return queueName
   * Search if the queueName is a key value => return the value
   * else return the queue passed on parameter
   */
  getQueueName(queueName: string) {
    const defaultValue = this.valueExtractor(`"${queueName}"`, queueName);

    if (this.valueExtractor(`application.amqp[0] ? true : false`, false)) {
      return this.valueExtractor(
        `application.amqp[${
          this.indexConfig
        }].queues['${queueName}'].queueName`,
        defaultValue
      );
    } else {
      return this.valueExtractor(
        `application.amqp.queues['${queueName}'].queueName`,
        defaultValue
      );
    }
  }

  /**
   * Convert content to buffer for send in queue
   */
  async contentToBuffer(content: any) {
    let data: any;

    if (content === undefined) {
      data = this.undefinedValue;
    } else if (content instanceof Buffer) {
      data = content;
    } else if (typeof content === 'string') {
      data = content;
    } else if (content instanceof Error) {
      data = JSON.stringify(content, Object.getOwnPropertyNames(content));
    } else {
      data = JSON.stringify(content);
    }

    if (this.gzipEnabled) {
      return new Promise<Buffer>(resolve => {
        gzip(new Buffer(data), (err, res) => {
          if (err) {
            resolve(undefined);
          } else {
            resolve(res);
          }
        });
      });
    }

    return Promise.resolve(new Buffer(data));
  }

  /**
   * Parse content in message
   */
  async parseContent(msg: Message) {
    let data: any;
    if (this.gzipEnabled) {
      data = await new Promise<any>(resolve => {
        gunzip(msg.content, (err, res) => {
          if (err) {
            resolve(msg.content.toString());
          } else {
            resolve(res.toString());
          }
        });
      });
    } else {
      data = msg.content.toString();
    }

    try {
      data = JSON.parse(data);
    } catch {}

    if (data === this.undefinedValue) {
      return undefined;
    }

    return data;
  }

  private constructListener(
    handlerMetadata: RabbitHandlerMetadata,
    parametersMetadata: ParameterMetadata[],
    controller: Controller
  ): ConsumerHandler {
    return async (msg: ConsumeMessage | null) => {
      if (msg === null) {
        return;
      }

      const args = await this.extractArgs(parametersMetadata, msg);
      await Promise.resolve(controller[handlerMetadata.key](...args));
      await this.channel.ack(msg);
    };
  }

  private constructConsumer(
    handlerMetadata: RabbitHandlerMetadata,
    parametersMetadata: ParameterMetadata[],
    controller: Controller
  ): ConsumerHandler {
    return async (msg: Message | null) => {
      if (msg === null) {
        return;
      }

      // catch when error amqp (untestable)
      /* istanbul ignore next */
      if (msg.properties.replyTo === undefined) {
        throw new Error(`replyTo is missing`);
      }

      const args = await this.extractArgs(parametersMetadata, msg);

      let response: any;
      let sendOptions: SendOptions;
      try {
        response = await Promise.resolve(
          controller[handlerMetadata.key](...args)
        );
        sendOptions = handlerMetadata.sendOptions || {};
      } catch (err) {
        response = err;
        sendOptions = handlerMetadata.sendOptionsError || {};
      }

      this.sendToQueueAck(msg.properties.replyTo, response, msg, {
        correlationId: msg.properties.correlationId,
        contentType: 'application/json',
        ...sendOptions,
      });
    };
  }

  private async extractArgs(
    params: ParameterMetadata[],
    msg: Message
  ): Promise<any[]> {
    const args = [];
    const content = await this.parseContent(msg);

    if (!params || !params.length) {
      return [content];
    }

    for (const item of params) {
      switch (item.type) {
        case PARAMETER_TYPE.MESSAGE:
          args[item.index] = msg;
          break;
        case PARAMETER_TYPE.CONTENT:
          args[item.index] = this.getParam(content, null, item);
          break;
        case PARAMETER_TYPE.PROPERTIES:
          args[item.index] = this.getParam(msg, 'properties', item);
          break;
        case PARAMETER_TYPE.FIELDS:
          args[item.index] = this.getParam(msg, 'fields', item);
          break;
        default:
          args[item.index] = msg;
          break; // response
      }
    }

    args.push(content);
    return args;
  }

  private getParam(
    source: any,
    paramType: string | null,
    itemParam: ParameterMetadata
  ) {
    const name = itemParam.parameterName;

    // get the param source
    let param = source;
    if (paramType !== null && source[paramType]) {
      param = source[paramType];
    }

    const res = param[name];

    if (res !== undefined) {
      return res;
    } else {
      return name === DEFAULT_PARAM_VALUE ? param : undefined;
    }
  }

  private getChannel(): ConfirmChannel | null {
    return (<any>this.channel)._channel;
  }
}
