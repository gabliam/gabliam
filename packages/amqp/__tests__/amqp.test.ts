import { AmqpPluginTest } from './amqp-plugin-test';
import {
  RabbitController,
  RabbitListener,
  RabbitConsumer,
  AmqpConnectionManager,
  CUnit,
} from '../src/index';
import {
  config,
  configWithName,
  configWith2Connection,
  configWithDuplicateConnection,
} from './conf';
import { AmqpConnection } from '../src/amqp-connection';
import { Gabliam } from '@gabliam/core';
import { Deferred } from './defered';
import * as sinon from 'sinon';

let appTest: AmqpPluginTest;
beforeEach(async () => {
  appTest = new AmqpPluginTest();
});

afterEach(async () => {
  try {
    await appTest.gab.stop();
  } catch (e) {}
  try {
    await appTest.destroy();
  } catch (e) {}
});

describe('Errors', () => {
  test('without config folder', async () => {
    await expect(appTest.gab.buildAndStart()).rejects.toMatchSnapshot();
  });

  test('with bad config', async () => {
    appTest.addConf('application.amqp', 'bad');
    await expect(appTest.gab.buildAndStart()).rejects.toMatchSnapshot();
  });

  test('with bad url', async () => {
    appTest.addConf('application.amqp', {
      url: 'amqp://bad',
      queues: {
        listener: {
          queueName: 'listenerTest',
          options: {
            durable: false,
          },
        },
      },
    });
    await expect(appTest.gab.buildAndStart()).rejects.toMatchSnapshot();
  });

  test('without queueName', async () => {
    appTest.addConf('application.amqp', {
      url: 'amqp://bad',
      queues: {
        queueTest: {
          name: 'lol',
        },
      },
    });
    await expect(appTest.gab.buildAndStart()).rejects.toMatchSnapshot();
  });
});

describe('Listener test', () => {
  [
    {
      name: 'with buffer string',
      content: new Buffer('buffertest'),
    },
    {
      name: 'with buffer object',
      content: new Buffer(JSON.stringify({ test: 'buffertest' })),
    },
    {
      name: 'with string',
      content: 'test',
    },
    {
      name: 'with object',
      content: { test: 'cool' },
    },
    {
      name: 'with undefined',
      content: undefined,
    },
  ].forEach(testCase => {
    test(testCase.name, async () => {
      const d = new Deferred();
      @RabbitController()
      class ControllerTest {
        @RabbitListener('listenerTest')
        listener(msg: any) {
          d.resolve(msg);
        }
      }

      const spy = sinon.spy(ControllerTest.prototype, 'listener');

      appTest.addConf('application.amqp', config);
      appTest.addClass(ControllerTest);
      await expect(appTest.gab.buildAndStart()).resolves.toBeInstanceOf(
        Gabliam
      );
      const connection = appTest.gab.container.get(AmqpConnection);
      connection.sendToQueue('listenerTest', testCase.content);
      const resp = await d.promise;
      expect(resp).toMatchSnapshot();
      expect(spy.callCount).toMatchSnapshot();
      expect(spy.args).toMatchSnapshot();
      spy.resetHistory();
    });
  });
});

describe('Consumer test', () => {
  [
    {
      name: 'with buffer string',
      content: new Buffer('buffertest'),
    },
    {
      name: 'with buffer object',
      content: new Buffer(JSON.stringify({ test: 'buffertest' })),
    },
    {
      name: 'with string',
      content: 'test',
    },
    {
      name: 'with object',
      content: { test: 'cool' },
    },
  ].forEach(testCase => {
    test(testCase.name, async () => {
      @RabbitController()
      class ControllerTest {
        @RabbitConsumer('consumerTest')
        consumer(msg: any) {
          return msg;
        }
      }

      const spy = sinon.spy(ControllerTest.prototype, 'consumer');

      appTest.addConf('application.amqp', config);
      appTest.addClass(ControllerTest);
      await expect(appTest.gab.buildAndStart()).resolves.toBeInstanceOf(
        Gabliam
      );
      const connection = appTest.gab.container.get(AmqpConnection);
      const resp = await connection.sendAndReceive(
        'consumerTest',
        testCase.content
      );
      expect(resp).toMatchSnapshot();
      expect(spy.callCount).toMatchSnapshot();
      expect(spy.args).toMatchSnapshot();
      spy.resetHistory();
    });
  });
});

test('consumer throw error', async () => {
  @RabbitController()
  class ControllerTest {
    @RabbitConsumer('consumerTest')
    consumer(msg: any) {
      throw new Error('Consumer error');
    }
  }

  const spy = sinon.spy(ControllerTest.prototype, 'consumer');

  appTest.addConf('application.amqp', config);
  appTest.addClass(ControllerTest);
  await expect(appTest.gab.buildAndStart()).resolves.toBeInstanceOf(Gabliam);
  const connection = appTest.gab.container.get(AmqpConnection);
  const resp = await connection.sendAndReceive('consumerTest', 'test');
  expect(resp).toMatchSnapshot<any>({
    stack: expect.any(String),
  });
  expect(spy.callCount).toMatchSnapshot();
  expect(spy.args).toMatchSnapshot();
  spy.resetHistory();
});

test('sendAndReceive timeout', async () => {
  appTest.addConf('application.amqp', config);
  await expect(appTest.gab.buildAndStart()).resolves.toBeInstanceOf(Gabliam);
  const connection = appTest.gab.container.get(AmqpConnection);
  await expect(
    connection.sendAndReceive('consumerTest', 'test', {}, 100)
  ).rejects.toMatchSnapshot();
});

test('bad queue', async () => {
  @RabbitController()
  class ControllerTest {
    @RabbitConsumer('consumerTestBad')
    consumer(msg: any) {
      throw new Error('Consumer error');
    }
  }

  appTest.addConf('application.amqp', config);
  appTest.addClass(ControllerTest);
  await expect(appTest.gab.buildAndStart()).rejects.toMatchSnapshot();
});

test('start 2 times', async () => {
  @RabbitController()
  class ControllerTest {
    @RabbitConsumer('consumerTest')
    consumer(msg: any) {
      return msg;
    }
  }

  const spy = sinon.spy(ControllerTest.prototype, 'consumer');

  appTest.addConf('application.amqp', config);
  appTest.addClass(ControllerTest);
  await expect(appTest.gab.buildAndStart()).resolves.toBeInstanceOf(Gabliam);
  await expect(appTest.startPlugins('AmqpPlugin'));
  const connection = appTest.gab.container.get(AmqpConnection);
  const resp = await connection.sendAndReceive(
    'consumerTest',
    'testCase.content'
  );
  expect(resp).toMatchSnapshot();
  expect(spy.callCount).toMatchSnapshot();
  expect(spy.args).toMatchSnapshot();
  spy.resetHistory();
});

test('must fail when getConnection not found', async () => {
  @RabbitController()
  class ControllerTest {
    @RabbitConsumer('consumerTest')
    consumer(msg: any) {
      return msg;
    }
  }

  appTest.addConf('application.amqp', config);
  appTest.addClass(ControllerTest);
  await expect(appTest.gab.buildAndStart()).resolves.toBeInstanceOf(Gabliam);
  await expect(appTest.startPlugins('AmqpPlugin'));
  const connection = appTest.gab.container.get(AmqpConnectionManager);
  await expect(() =>
    connection.getConnection('notfound')
  ).toThrowErrorMatchingSnapshot();
});

test('with one connection with name and entity without cunit', async () => {
  @RabbitController()
  class ControllerTest {
    @RabbitConsumer('consumerTest')
    consumer(msg: any) {
      return msg;
    }
  }

  const spy = sinon.spy(ControllerTest.prototype, 'consumer');

  appTest.addConf('application.amqp', configWithName);
  appTest.addClass(ControllerTest);
  await expect(appTest.gab.buildAndStart()).resolves.toBeInstanceOf(Gabliam);
  await expect(appTest.startPlugins('AmqpPlugin'));
  const connection = appTest.gab.container.get(AmqpConnection);
  const resp = await connection.sendAndReceive(
    'consumerTest',
    'testCase.content'
  );
  expect(resp).toMatchSnapshot();
  expect(spy.callCount).toMatchSnapshot();
  expect(spy.args).toMatchSnapshot();
  spy.resetHistory();
});

test('with config 2 database', async () => {
  @CUnit('connection1')
  @RabbitController()
  class ControllerTestConnection1 {
    @RabbitConsumer('consummer')
    consumer(msg: any) {
      return 'connection1' + msg;
    }
  }

  @CUnit('connection2')
  @RabbitController()
  class ControllerTestConnection2 {
    @RabbitConsumer('consummer')
    consumer(msg: any) {
      return 'connection2' + msg;
    }
  }

  appTest.addConf('application.amqp', configWith2Connection);
  appTest.addClass(ControllerTestConnection1);
  appTest.addClass(ControllerTestConnection2);

  await expect(appTest.gab.buildAndStart()).resolves.toBeInstanceOf(Gabliam);

  const connectionManager = appTest.gab.container.get(AmqpConnectionManager);

  const connection1 = connectionManager.getConnection('connection1');

  const resp = await connection1.sendAndReceive('consummer', 'testCase');
  expect(resp).toMatchSnapshot();

  const connection2 = connectionManager.getConnection('connection1');

  const resp2 = await connection2.sendAndReceive('consummer', 'testCase');
  expect(resp2).toMatchSnapshot();
});

test('must fail when CUnit not found', async () => {
  appTest.addConf('application.amqp', config);

  @CUnit('bad')
  @RabbitController()
  class ControllerTest {
    @RabbitConsumer('consumerTest')
    consumer(msg: any) {
      return msg;
    }
  }
  appTest.addClass(ControllerTest);
  await expect(appTest.gab.buildAndStart()).rejects.toMatchSnapshot();
});

test('must fail with duplicate connections', async () => {
  appTest.addConf('application.amqp', configWithDuplicateConnection);
  await expect(appTest.gab.buildAndStart()).rejects.toMatchSnapshot();
});
