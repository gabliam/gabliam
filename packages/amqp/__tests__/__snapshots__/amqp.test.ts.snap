// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Consumer test with buffer object 1`] = `
Object {
  "test": "buffertest",
}
`;

exports[`Consumer test with buffer object 2`] = `1`;

exports[`Consumer test with buffer object 3`] = `
Array [
  Array [
    Object {
      "test": "buffertest",
    },
  ],
]
`;

exports[`Consumer test with buffer string 1`] = `"buffertest"`;

exports[`Consumer test with buffer string 2`] = `1`;

exports[`Consumer test with buffer string 3`] = `
Array [
  Array [
    "buffertest",
  ],
]
`;

exports[`Consumer test with object 1`] = `
Object {
  "test": "cool",
}
`;

exports[`Consumer test with object 2`] = `1`;

exports[`Consumer test with object 3`] = `
Array [
  Array [
    Object {
      "test": "cool",
    },
  ],
]
`;

exports[`Consumer test with string 1`] = `"test"`;

exports[`Consumer test with string 2`] = `1`;

exports[`Consumer test with string 3`] = `
Array [
  Array [
    "test",
  ],
]
`;

exports[`Errors with bad config 1`] = `[Error: Error for 'application.amqp' value{"isJoi":true,"name":"ValidationError","details":[{"message":"/"value/" must be an object","path":[],"type":"object.base","context":{"value":"bad","label":"value"}},{"message":"/"value/" must be an array","path":[],"type":"array.base","context":{"label":"value"}}],"_object":"bad"}]`;

exports[`Errors with bad url 1`] = `[Error: getaddrinfo ENOTFOUND bad bad:5672]`;

exports[`Errors without config folder 1`] = `[Error: AmqpPluginConfig AmqpConfig is mandatory]`;

exports[`Errors without queueName 1`] = `[Error: Error for 'application.amqp' value{"isJoi":true,"name":"ValidationError","details":[{"message":"/"queueName/" is required","path":["queues","queueTest","queueName"],"type":"any.required","context":{"key":"queueName","label":"queueName"}},{"message":"/"name/" is not allowed","path":["queues","queueTest","name"],"type":"object.allowUnknown","context":{"child":"name","value":"lol","key":"name","label":"name"}},{"message":"/"value/" must be an array","path":[],"type":"array.base","context":{"label":"value"}}],"_object":{"url":"amqp://bad","queues":{"queueTest":{"name":"lol"}}}}]`;

exports[`Listener test with buffer object 1`] = `
Object {
  "test": "buffertest",
}
`;

exports[`Listener test with buffer object 2`] = `1`;

exports[`Listener test with buffer object 3`] = `
Array [
  Array [
    Object {
      "test": "buffertest",
    },
  ],
]
`;

exports[`Listener test with buffer string 1`] = `"buffertest"`;

exports[`Listener test with buffer string 2`] = `1`;

exports[`Listener test with buffer string 3`] = `
Array [
  Array [
    "buffertest",
  ],
]
`;

exports[`Listener test with object 1`] = `
Object {
  "test": "cool",
}
`;

exports[`Listener test with object 2`] = `1`;

exports[`Listener test with object 3`] = `
Array [
  Array [
    Object {
      "test": "cool",
    },
  ],
]
`;

exports[`Listener test with string 1`] = `"test"`;

exports[`Listener test with string 2`] = `1`;

exports[`Listener test with string 3`] = `
Array [
  Array [
    "test",
  ],
]
`;

exports[`Listener test with undefined 1`] = `undefined`;

exports[`Listener test with undefined 2`] = `1`;

exports[`Listener test with undefined 3`] = `
Array [
  Array [
    undefined,
  ],
]
`;

exports[`bad queue 1`] = `[Error: queue "consumerTestBad" doesn't exist]`;

exports[`consumer throw error 1`] = `
Object {
  "message": "Consumer error",
  "stack": Any<String>,
}
`;

exports[`consumer throw error 2`] = `1`;

exports[`consumer throw error 3`] = `
Array [
  Array [
    "test",
  ],
]
`;

exports[`must fail when CUnit not found 1`] = `[Error: Connection bad not found]`;

exports[`must fail when getConnection not found 1`] = `"Connection notfound not found"`;

exports[`must fail with duplicate connections 1`] = `[Error: Duplicate connection connection1]`;

exports[`sendAndReceive timeout 1`] = `[Error: operation timed out]`;

exports[`start 2 times 1`] = `"testCase.content"`;

exports[`start 2 times 2`] = `1`;

exports[`start 2 times 3`] = `
Array [
  Array [
    "testCase.content",
  ],
]
`;

exports[`with config 2 database 1`] = `"connection1testCase"`;

exports[`with config 2 database 2`] = `"connection1testCase"`;

exports[`with one connection with name and entity without cunit 1`] = `"testCase.content"`;

exports[`with one connection with name and entity without cunit 2`] = `1`;

exports[`with one connection with name and entity without cunit 3`] = `
Array [
  Array [
    "testCase.content",
  ],
]
`;
